### PR TITLE
Ames - packets, diagrams, serf and king

### PR DESCRIPTION
**Pull request by syzygyzer**
This is the second of several Ames PR's I've been working on (the first being #943), which are mostly being done now for the sake of the security audit but of course the information is widely useful. It is mostly an elaboration on what was already summarized at the top.

This adds two major sections:
- Packets: what format they come in, how they are encrypted, how they are serialized, how messages are packetized, how they are acked, and how they are relayed.
- The Serf and the King: The Serf does most of the grunt work for Ames, but the King has important responsibilities too.

Each section is accompanied by a diagram that helps to describe their content.

There are still some important things to come. A discussion of nacks and failures in Ames in general, how Ames and Jael work together to handle keys, and what specifically constitutes an Arvo event and how that relates to flows. If there are other parts of Ames you'd like documented, now is the time to mention it.

### CODE REVIEW

**Comment by syzygyzer** [comment]
I've added content on flows, nacks, and naxplanations, and reworded a few other things as appropriate.

**Comment by syzygyzer** [comment]
pinging @belisarius222

File: tutorials/arvo/ames.md [hunk 1]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+body.
+
+A nack will be accompanied by a naxplanation, which is a third type of message
+separate from `%plea`s and `%boon`s. Ames won't give the vane that requested the
+initial `%plea` a nack until it also receives the naxplanation, which it will send to
+the vane as a `%done` gift. Naxplanations may only be acked, never nacked.
+
+#### (N)ack packets
+
+The contents of a given (n)ack packet are deterministic. The `content` (i.e. the
```
[the remark this answers is not in the document]
Comment (belisarius222) [reply tutorials/arvo/ames.md:233]: Maybe a more precise way of stating what you mean here would be to say something like:
```
A fragment ack's contents (before encryption) are a pure function of the following data:
[our-life her-life ...]

This means all re-sends of an ack packet will be bitwise identical to each other, unless one of the peers changes its encryption keys.
```

The same is true for message acks and message fragment packets, too.  The unifying concept here, which I'm not sure is really explained anywhere, is that before encryption (which introduces a dependency on the sender's and receiver's life numbers) and relaying (which can modify the `origin` field), every Ames packet is a pure function of `[sndr rcvr flow-number message-number packet-id]`, where `packet-id` varies by packet type, but is either `[%message-fragment fragment-number]`, `[%fragment-ack fragment-number]`, or `[%message-ack nack=(unit naxplanation)]`.

File: tutorials/arvo/ames.md [hunk 2]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+```
+Each datum in this noun is an atom with the aura `@ud` or an aura that nests
+under `@ud`.
+
+Here, `our-life` refers to the [`life`](@/docs/glossary/breach), or revision
+number, of the acking ship's networking keys, and `her-life` is the `life` of
+the ack-receiving ship's networking keys. `bone` is an opaque number identifying the flow, `message-num` denotes the number of the
+message in the flow identified by `bone`, and
+`fragment-num` denotes the number of the fragment of the message identified by
+`message-num` being acked.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:248]: The "being acked" at the end of the sentence is hard to parse.  Maybe split this into multiple sentences?

File: tutorials/arvo/ames.md [hunk 3]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+under `@ud`.
+
+Here, `our-life` refers to the [`life`](@/docs/glossary/breach), or revision
+number, of the acking ship's networking keys, and `her-life` is the `life` of
+the ack-receiving ship's networking keys. `bone` is an opaque number identifying the flow, `message-num` denotes the number of the
+message in the flow identified by `bone`, and
+`fragment-num` denotes the number of the fragment of the message identified by
+`message-num` being acked. 
+
+A message (n)ack is an extended fragment ack that is obtained by encrypting the
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:250]: > extended fragment ack
Not really.  It's a different kind of ack, with different contents.

File: tutorials/arvo/ames.md [hunk 4]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+`+jam` of the following noun:
+
+```hoon
+[our-life her-life bone message-num fragment-num ok=? lag=@dr]
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:254]: There's no `fragment-num` in a message ack.  The ack packet attests to having processed the whole message, so it doesn't need to refer to any particular fragment.

File: tutorials/arvo/ames.md [hunk 5]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+
+Flows are asymmetric in that one ship is always the sender of
+`%plea`s and the other is always the sender of `%boon`s on a given flow. Thus ships will
+typically have multiple flows between them.
+
+The sequence of `%plea`s in a flow is totally ordered, though
+the packets which make them up need not be. `%boon`s are totally ordered in the
+same fashion, but there is not a coordinated ordering between `%plea`s and
+`%boon`s in a given flow beyond the implicit one arising from the fact that a
+`%boon` is always a response to a `%plea`.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:274]: I think it would be good to clarify here that each `%plea` can give rise to zero, one, or many `%boon`s.

Comment (belisarius222) [reply tutorials/arvo/ames.md:274]: You could also say here that to create a new flow, a ship sends a `%plea` to another ship.  Only the creator of a flow can send a `%plea` on it, and only the other party can send a `%boon` on it.

File: tutorials/arvo/ames.md [hunk 6]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+
+Inside of Ames, each flow has four sequential opaque `@ud`s
+called bones that are unique to that flow. Thus the flow itself is often
+referred to by its first bone. The set of types of packets is partitioned into
+four subsets, each of which is assigned a bone. Each bone is a one-way street
+for packets to travel along, so e.g. acks to packets making up a `%plea` are
+sent along different bones.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:281]: They're not sent on different bones from each other; all acks to `%plea`s on a given flow will be on the same bone.

File: tutorials/arvo/ames.md [hunk 7]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+The sequence of `%plea`s in a flow is totally ordered, though
+the packets which make them up need not be. `%boon`s are totally ordered in the
+same fashion, but there is not a coordinated ordering between `%plea`s and
+`%boon`s in a given flow beyond the implicit one arising from the fact that a
+`%boon` is always a response to a `%plea`.
+
+Inside of Ames, each flow has four sequential opaque `@ud`s
+called bones that are unique to that flow. Thus the flow itself is often
+referred to by its first bone. The set of types of packets is partitioned into
+four subsets, each of which is assigned a bone. Each bone is a one-way street
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:279]: > The set of types of packets is partitioned into four subsets, each of which is assigned a bone.

I find this sentence hard to understand, and this concept is explained in more detail later.  I'd remove the sentence.

File: tutorials/arvo/ames.md [hunk 8]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+`~bacbel-tagfeg` as the `%plea` sender. Then bones 12-15 are associated with the
+flow, and the types of packets for each bone are:
+
+ - bone 12, `%plea`s and acks to `~worwel-sipnum`,
+ - bone 13, `%boon`s and (n)acks to `~bacbel-tagfeg`,
+ - bone 14, acks of naxplanations to `~worwel-sipnum`,
+ - bone 15, naxplanations to `~bacbel-tagfeg`.
+
+Each bone is handled separately by congestion control, and this is the reason
+for their segregation. For example, say a two packet `%plea` is sent on bone 12,
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:293]: "two packet" -> "two-packet"

File: tutorials/arvo/ames.md [hunk 9]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+We give an example of such a partition. Let flow 12 be a flow between `~bacbel-tagfeg` and `~worwel-sipnum` with
+`~bacbel-tagfeg` as the `%plea` sender. Then bones 12-15 are associated with the
+flow, and the types of packets for each bone are:
+
+ - bone 12, `%plea`s and acks to `~worwel-sipnum`,
+ - bone 13, `%boon`s and (n)acks to `~bacbel-tagfeg`,
+ - bone 14, acks of naxplanations to `~worwel-sipnum`,
+ - bone 15, naxplanations to `~bacbel-tagfeg`.
+
+Each bone is handled separately by congestion control, and this is the reason
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:292]: Separate congestion control is one reason to separate bones.  There are two others.

It's important that I be able to start a flow with you without risk of race conditions.  That is, if I try to start a flow with you at the same time you try to start a flow with me, when we receive each other's packets, that shouldn't cause a conflict.  Flipping the last bit of the bone based on whether a packet is an ack or a message fragment allows for me to create flow 8 with you without that ever coming into conflict with the flow 8 that you created with me.

The reason to have four bones per flow, not just two, is that naxplanation packets shouldn't be able to conflict with happy-path packets.  Since naxplanations are generated dynamically in response to hearing messages, if they were  on the same bones as happy-path messages, they would have to occupy new numbers within those flows.  Maybe it would be possible to get that to work, but it was easier to be confident in the correctness of the design with naxplanations on their own bones that can't conflict with the happy path.

File: tutorials/arvo/ames.md [hunk 10]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+   is the `+mug` of the body, and the sender and receiver ship types are `01`,
+   denoting that the sender and receiver are planets. The remainder of the
+   packet is the body.
+ - The body of this packet will be the `@p` of the
+sender `~bacbel-tagfeb`, followed by the `@p` of the receiver `~worwel-sipnum`,
+followed by the payload.
+ - The payload of this packet will be the `+jam` of `[origin content]`. `origin`
+for this initial packet will be `~`, which implies that the packet is
+originating from the sender. `content` is the `+jam` of the message, encrypted
+using a Diffie-Hellman symmetric key obtained with the private key of
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:355]: I think this phrasing is a bit ambiguous.  The symmetric key is obtained using a Diffie-Hellman, once, *and then* the encryption of each packet is performed with that symmetric key.

File: tutorials/arvo/ames.md [hunk 11]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+`~worwel-sipnum`. First, it [`+cue`s](@/docs/reference/library/2p.md#cue)
+(deserializes) the payload and changes the `origin` to the IP and port of
+`~bacbel-tagfeb`. Then it `+jam`s `[origin content]` to form a new payload, and
+`+mug`s that payload to get a new checksum. It replaces the old payload and
+checksum with the new, and then sends the packet along to `~worwel-sipnum`.
+
+Once `~worwel-sipnum` processes the packet, it will know the IP and port of
+`~bacbel-tagfeb` since `~zod` included it when it forwarded the packet and so
+will send an ack packet directly to `~bacbel-tagfeb`. Communication between the
+two ships will now be direct until one of them changes their IP address, port, or networking keys.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:369]: NATs and firewalls can prevent a direct peer-to-peer connection, even if the two ships know each other's IP and port.  In that case, a star or galaxy will continue to relay packets.

File: tutorials/arvo/ames.md [hunk 12]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+
+
+## The Serf and the King
+
+Urbit's functionality is split between the two binaries `urbit-worker` (sometimes
+called the Serf) and `urbit-king` (sometimes called the King). This division of
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:375]: I think you mean `urbit-daemon`, not `urbit-king`.

File: tutorials/arvo/ames.md [hunk 13]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+labor is currently not well-documented outside of the [Vere documents](@/docs/tutorials/vere/_index.md), but we summarize it here.
+
+In short, the Serf is the Nock runtime and so keeps track of the current state
+of Arvo as a Nock noun and updates the state by `%poke`ing it with nouns, and
+then informs the King of the new state. The King manages snapshots of the Arvo
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:380]: It's actually the worker, not the daemon, that manages Arvo snapshots, since that's the process that holds the Arvo state in memory.

File: tutorials/arvo/ames.md [hunk 14]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+state and handles I/O with Unix, among other things. The Serf only ever talks to
+the King, while the King talks with both the Serf and Unix.
+
+### Ames
+
+The King has several submodules, one of them being an Ames I/O submodule. This
+submodule is responsible for wrapping outgoing Ames packets as UDP packets, and
+looking at incoming UDP packets handed to it by Unix to ensure that they are
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:388]: It doesn't do much validation; it mostly turns the packets into Ames packets and sends them to the worker to be processed.

File: tutorials/arvo/ames.md [hunk 15]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+valid Ames packets. It also maintains an incoming and outgoing packet queue.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:389]: Outgoing packets are essentially fired off immediately, not really enqueued.

File: tutorials/arvo/ames.md [hunk 16]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+</div>
+
+Ames, as a part of Arvo, handles `+jam`ming, packetizing, encryption, and
+forming Ames packets. Once it is ready to send an Ames packet, it `%give`s to
+Unix a `%send` `gift` containing that packet. This will be a Nock noun
+containing the `@tas` `%send` as well as the serialized packet.
+
+"Unix", in this case, is actually the King. The King receives the `%send`
+instruction, wraps the packet contained within as a UDP packet, and queues the
+UDP packet contained. Once it is time to send it (which will be determined by
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:406]: It's the Ames vane that performs congestion control logic, not the daemon process.  Once the daemon process is informed it's supposed to send a UDP packet, it fires it off without question.

File: tutorials/arvo/ames.md [hunk 17]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+congestion control), it will go ahead and forward the packet to Unix's
+networking interface, at which point it is no longer controlled by any part of Urbit.
+
+Now the receiving King is handed a UDP packet by Unix. The King removes the UDP
+wrapped and checks to see that it has received a valid Ames packet. If so, it
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:411]: "wrapped" -> "wrapper"

Comment (belisarius222) [reply tutorials/arvo/ames.md:411]: The daemon also doesn't fully check packet validity before sending it to be processed by the Ames vane.

File: tutorials/arvo/ames.md [hunk 18]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+`+cue`s the payload and wraps it as a `%hear` `note` and passes it to the Serf.
```
Comment (belisarius222) [inline tutorials/arvo/ames.md:412]: The daemon doesn't `+cue` the payload; it jams the lane on which it heard the packet, but other than that, it delivers the packet to the Ames vane as an atom by coping the bytes it heard on its UDP port.

**Review by belisarius222 — CHANGES_REQUESTED** [review]

**Comment by syzygyzer** [comment]
@belisarius222 I plan to finish addressing your review today or tomorrow. I am wondering though, with the upcoming vere release, does the serialization steps described here change? I discuss two serialization steps, but I remember during the infra meeting last week (I think) it was mentioned that there were even more than that?

"- Improved event loop performance by removing unnecessary serialization."

**Comment by belisarius222** [comment]
That's an unrelated issue, I believe. We might change the packet format in other ways when we breach, but that isn't set in stone yet.

—
~rovnys-ricfer
https://urbit.org

File: tutorials/arvo/ames.md [hunk 19]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+
+#### Nacks
+
+A nack indicates a negative acknowledgement to a `%plea`, meaning that the
+requested action was not performed.
+
+`%boon`s and naxplanations are never nacked.
+Individual packets are never nacked, only complete `%plea` messages. This works
+because only complete packets ever make it to Ames, as malformed
+packets will be discarded by the Ames I/O module in the
```
[the remark this answers is not in the document]
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR216

File: tutorials/arvo/ames.md [hunk 20]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+[King](#the-serf-and-the-king). Therefore there is no need for Ames to do
+something like nack a packet whose checksum does not match the `+mug` of the
```
[the remark this answers is not in the document]
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR219

File: tutorials/arvo/ames.md [hunk 21]
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
...
+body.
+
+A nack will be accompanied by a naxplanation, which is a third type of message
```
[the remark this answers is not in the document]
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR225

File: tutorials/arvo/ames.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR234-R240

File: tutorials/arvo/ames.md (hunk 9 above)
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/d42f0a3aee11fd2b410448332da010a1a880ac23#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR302-R320

How's this?

Comment (syzygyzer) [reply tutorials/arvo/ames.md]: Whoops, somehow started a new review for that comment. I'm gonna leave it just to ensure I don't fuck something else up.

File: tutorials/arvo/ames.md (hunk 10 above)
Comment (syzygyzer) [reply tutorials/arvo/ames.md]: https://github.com/urbit/docs/pull/957/commits/3f725698e167a6d4ff1cf726afbe515cb322a755#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daL351-R366

File: tutorials/arvo/ames.md (hunk 9 above)
Comment (belisarius222) [reply tutorials/arvo/ames.md]: LGTM

File: tutorials/arvo/ames.md (hunk 13 above)
Comment (syzygyzer) [reply tutorials/arvo/ames.md:390]: I'm a bit confused - I thought the daemon and the worker were the same thing? Maybe I should be more clear of what is meant by manage - does the King save the snapshots to disk since it handles I/O? Or does the daemon (or worker? whichever you mean) actually also talk to Unix in order to save snapshots?

**Comment by syzygyzer** [comment]
@belisarius222 Alright I'm finished with this round of revisions. I was trying to handle the review with emacs, but I seemed to have messed something up and switched to just using the GH website partway through, so it got a little fucky, but for comments that were easy to resolve I just closed them and for ones I want you to look at I linked. Hopefully that does not end up being a PITA.

**Comment by belisarius222** [comment]
No, they're not the same; the daemon is the I/O process and the worker is the nock interpreter process.  The daemon writes events to disk, and the worker writes snapshots to disk.

—
~rovnys-ricfer
https://urbit.org

**Comment by syzygyzer** [comment]
What did you mean by "I think you mean `urbit-daemon`, not `urbit-king`" then? If the daemon is I/O process (i.e. the King), then what is the `urbit-king` binary if it is not the daemon?

**Comment by syzygyzer** [comment]
Ok I'm not sure how this edit ended up in this commit... I think I must have done something bad with this repo, or my emacs, since this was supposed to just be me changing the names around for the binaries again.
